File: src/lib/estimators/bocpd.ts
// Bayesian Online Change-Point Detection (Adams & MacKay 2007).
//
// Student-t posterior predictive from a Normal-Inverse-Gamma conjugate prior
// on (mean, variance) of Gaussian observations. With constant hazard H the
// posterior P(r_t = 0) ≡ H — not a useful change-point score — so we expose:
//
//   - runLengthPosterior  (T × T+1)
//   - mapRunLength        per-step argmax
//   - newRunProb          P(r_t < cpWindow): the trace to threshold
//   - predictiveMean / logEvidence
//
// Ported from research/estimators/bocpd.py; validated against the same
// reference fixture (src/lib/__tests__/estimators/fixtures/bocpd_reference.json).

export interface BocpdOptions {
  hazard?: number;
  mu0?: number;
  kappa0?: number;
  alpha0?: number;
  beta0?: number;
  cpWindow?: number;
}

export interface BocpdResult {
  runLengthPosterior: number[][];
  mapRunLength: number[];
  newRunProb: number[];
  predictiveMean: number[];
  logEvidence: number[];
  cpWindow: number;
}

// Lanczos approximation to log Γ, accurate to ~1e-13 on ℝ⁺.
const LANCZOS_G = 7;
const LANCZOS_C = [
  0.99999999999980993, 676.5203681218851, -1259.1392167224028,
  771.32342877765313, -176.61502916214059, 12.507343278686905,
  -0.13857109526572012, 9.9843695780195716e-6, 1.5056327351493116e-7,
];

export function lgamma(x: number): number {
  if (x < 0.5) {
    // Reflection: Γ(x)Γ(1-x) = π / sin(πx)
    return Math.log(Math.PI / Math.sin(Math.PI * x)) - lgamma(1 - x);
  }
  x -= 1;
  let a = LANCZOS_C[0];
  const t = x + LANCZOS_G + 0.5;
  for (let i = 1; i < LANCZOS_C.length; i++) a += LANCZOS_C[i] / (x + i);
  return 0.5 * Math.log(2 * Math.PI) + (x + 0.5) * Math.log(t) - t + Math.log(a);
}

export function logsumexp(a: number[]): number {
  let m = -Infinity;
  for (const v of a) if (v > m) m = v;
  if (!Number.isFinite(m)) return m;
  let s = 0;
  for (const v of a) s += Math.exp(v - m);
  return m + Math.log(s);
}

// Student-t log-density at scalar x with df, loc, scale.
function studentTLogpdf(x: number, df: number, loc: number, scale: number): number {
  const z = (x - loc) / scale;
  return (
    lgamma((df + 1) / 2) -
    lgamma(df / 2) -
    0.5 * Math.log(df * Math.PI) -
    Math.log(scale) -
    ((df + 1) / 2) * Math.log1p((z * z) / df)
  );
}

export function bocpd(x: ArrayLike<number>, opts: BocpdOptions = {}): BocpdResult {
  const hazard = opts.hazard ?? 1 / 250;
  const mu0 = opts.mu0 ?? 0;
  const kappa0 = opts.kappa0 ?? 1;
  const alpha0 = opts.alpha0 ?? 1;
  const beta0 = opts.beta0 ?? 1;
  const cpWindow = opts.cpWindow ?? 20;

  const T = x.length;
  const logH = Math.log(hazard);
  const log1mH = Math.log1p(-hazard);

  // log_R[t][r] in log-domain, shape (T+1, T+1), init to -Infinity.
  const logR: number[][] = new Array(T + 1);
  for (let t = 0; t <= T; t++) logR[t] = new Array(T + 1).fill(-Infinity);
  logR[0][0] = 0;

  let mu = [mu0];
  let kappa = [kappa0];
  let alpha = [alpha0];
  let beta = [beta0];

  const predictiveMean = new Array<number>(T).fill(0);
  const logEvidence = new Array<number>(T).fill(0);

  for (let t = 0; t < T; t++) {
    const xt = x[t];
    const nr = mu.length; // = t + 1
    const logPred = new Array<number>(nr);
    for (let r = 0; r < nr; r++) {
      const df = 2 * alpha[r];
      const scale = Math.sqrt((beta[r] * (kappa[r] + 1)) / (alpha[r] * kappa[r]));
      logPred[r] = studentTLogpdf(xt, df, mu[r], scale);
    }

    const prev = logR[t].slice(0, nr);
    // joint(r_t = r+1) = prev[r] + logPred[r] + log(1 - h)
    const logGrowth = new Array<number>(nr);
    const combined = new Array<number>(nr);
    for (let r = 0; r < nr; r++) {
      logGrowth[r] = prev[r] + logPred[r] + log1mH;
      combined[r] = prev[r] + logPred[r];
    }
    const logCp = logsumexp(combined) + logH;

    logR[t + 1][0] = logCp;
    for (let r = 0; r < nr; r++) logR[t + 1][r + 1] = logGrowth[r];

    // log evidence at step t = logsumexp over row 0..t+1
    const row = logR[t + 1].slice(0, nr + 1);
    const logEv = logsumexp(row);
    logEvidence[t] = logEv;

    // Normalize row to posterior.
    for (let r = 0; r <= nr; r++) logR[t + 1][r] -= logEv;

    // Predictive mean weighted by run-length posterior; mu extended with prior.
    const muExt = [mu0, ...mu];
    let pm = 0;
    for (let r = 0; r <= nr; r++) pm += Math.exp(logR[t + 1][r]) * muExt[r];
    predictiveMean[t] = pm;

    // NIG sufficient-statistic updates (Murphy 2007 §3).
    const kappaNew = new Array<number>(nr);
    const muNew = new Array<number>(nr);
    const alphaNew = new Array<number>(nr);
    const betaNew = new Array<number>(nr);
    for (let r = 0; r < nr; r++) {
      kappaNew[r] = kappa[r] + 1;
      muNew[r] = (kappa[r] * mu[r] + xt) / kappaNew[r];
      alphaNew[r] = alpha[r] + 0.5;
      betaNew[r] = beta[r] + (kappa[r] * (xt - mu[r]) * (xt - mu[r])) / (2 * kappaNew[r]);
    }
    mu = [mu0, ...muNew];
    kappa = [kappa0, ...kappaNew];
    alpha = [alpha0, ...alphaNew];
    beta = [beta0, ...betaNew];
  }

  const runLengthPosterior: number[][] = new Array(T);
  const mapRunLength = new Array<number>(T).fill(0);
  const newRunProb = new Array<number>(T).fill(0);
  for (let t = 0; t < T; t++) {
    const row = new Array<number>(T + 1);
    let best = -Infinity;
    let bestIdx = 0;
    let winSum = 0;
    for (let r = 0; r <= T; r++) {
      const p = Math.exp(logR[t + 1][r]);
      row[r] = p;
      if (p > best) {
        best = p;
        bestIdx = r;
      }
      if (r < cpWindow) winSum += p;
    }
    runLengthPosterior[t] = row;
    mapRunLength[t] = bestIdx;
    newRunProb[t] = winSum;
  }

  return {
    runLengthPosterior,
    mapRunLength,
    newRunProb,
    predictiveMean,
    logEvidence,
    cpWindow,
  };
}

export function detectChangepointsFromMap(
  mapRunLength: ArrayLike<number>,
  minDrop = 10,
  refractory = 10,
): number[] {
  const out: number[] = [];
  let last = -refractory - 1;
  for (let i = 1; i < mapRunLength.length; i++) {
    if (mapRunLength[i] - mapRunLength[i - 1] < -minDrop) {
      if (i - last >= refractory) {
        out.push(i);
        last = i;
      }
    }
  }
  return out;
}

File: research/estimators/bocpd.py
"""Bayesian Online Change-Point Detection.

Adams & MacKay 2007 (https://arxiv.org/abs/0710.3742) with a Student-t
predictive distribution induced by a Normal-Inverse-Gamma conjugate prior
on (mean, variance) of Gaussian observations.

Change-point signal
-------------------
With constant hazard H, the posterior P(r_t = 0 | x_{1:t}) is identically
H (a normalization tautology), so it is *not* a useful change-point score.
The informative signal is the collapse of the run-length posterior toward
small r following a change-point. We expose:

- `run_length_posterior`: the full P(r_t = k | x_{1:t}) matrix.
- `map_run_length`: the MAP trajectory. At change-points, this either
  drops to 0 / small values or continues to track the new run from zero.
- `new_run_prob`: P(r_t < cp_window | x_{1:t}) — mass concentrated in
  runs younger than `cp_window` observations. High at change-points,
  near-zero during stationary stretches. This is the trace to threshold.

The NIG → Student-t predictive follows Murphy 2007 §3.
"""
from __future__ import annotations

from dataclasses import dataclass

import numpy as np
from scipy import stats


@dataclass
class BocpdResult:
    run_length_posterior: np.ndarray  # (T, T+1)
    map_run_length: np.ndarray  # (T,)
    new_run_prob: np.ndarray  # (T,) mass in r < cp_window
    predictive_mean: np.ndarray  # (T,)
    log_evidence: np.ndarray  # (T,) log P(x_t | x_{1:t-1})
    cp_window: int


def _student_t_logpdf(x: float, mu: np.ndarray, kappa: np.ndarray,
                      alpha: np.ndarray, beta: np.ndarray) -> np.ndarray:
    """Log posterior-predictive density under NIG(mu, kappa, alpha, beta).

    df    = 2α
    loc   = μ
    scale = sqrt(β(κ+1)/(ακ))
    """
    df = 2.0 * alpha
    scale = np.sqrt(beta * (kappa + 1.0) / (alpha * kappa))
    return stats.t.logpdf(x, df=df, loc=mu, scale=scale)


def bocpd(
    x: np.ndarray,
    hazard: float = 1.0 / 250.0,
    mu0: float = 0.0,
    kappa0: float = 1.0,
    alpha0: float = 1.0,
    beta0: float = 1.0,
    cp_window: int = 20,
) -> BocpdResult:
    """Run Adams & MacKay BOCPD on 1-D data `x`.

    Parameters
    ----------
    x : (T,) array of observations.
    hazard : constant hazard H ∈ (0, 1); higher = more frequent CPs expected.
    mu0, kappa0, alpha0, beta0 : NIG prior hyperparameters (Murphy 2007).
    cp_window : run-length window whose cumulative mass becomes the CP score.
        3 is a safe default; larger values smooth the score.
    """
    x = np.asarray(x, dtype=float).ravel()
    T = x.shape[0]

    # Log-domain run-length posterior for numerical stability on long series.
    log_R = np.full((T + 1, T + 1), -np.inf)
    log_R[0, 0] = 0.0

    mu = np.array([mu0])
    kappa = np.array([kappa0])
    alpha = np.array([alpha0])
    beta = np.array([beta0])

    predictive_mean = np.zeros(T)
    log_evidence = np.zeros(T)
    log_h = np.log(hazard)
    log_1mh = np.log1p(-hazard)

    for t in range(T):
        xt = x[t]
        log_pred = _student_t_logpdf(xt, mu, kappa, alpha, beta)

        prev = log_R[t, : t + 1]
        # joint log P(r_t = r+1, x_{1:t}) = log_R[t, r] + log_pred[r] + log(1-h)
        log_growth = prev + log_pred + log_1mh
        # joint log P(r_t = 0, x_{1:t}) = logsumexp_r(log_R[t, r] + log_pred[r]) + log(h)
        log_cp = _logsumexp(prev + log_pred) + log_h

        log_R[t + 1, 0] = log_cp
        log_R[t + 1, 1 : t + 2] = log_growth

        # Log evidence for this step = logsumexp over the joint row.
        log_ev = _logsumexp(log_R[t + 1, : t + 2])
        log_evidence[t] = log_ev

        # Normalize to posterior.
        log_R[t + 1, : t + 2] -= log_ev

        # Predictive mean weighted by posterior run-length.
        weights = np.exp(log_R[t + 1, : t + 2])
        mu_ext = np.concatenate([[mu0], mu])
        predictive_mean[t] = np.sum(weights * mu_ext)

        # NIG sufficient-statistic updates (Murphy 2007 §3).
        kappa_new = kappa + 1.0
        mu_new = (kappa * mu + xt) / kappa_new
        alpha_new = alpha + 0.5
        beta_new = beta + kappa * (xt - mu) ** 2 / (2.0 * kappa_new)

        mu = np.concatenate([[mu0], mu_new])
        kappa = np.concatenate([[kappa0], kappa_new])
        alpha = np.concatenate([[alpha0], alpha_new])
        beta = np.concatenate([[beta0], beta_new])

    run_length_posterior = np.exp(log_R[1 : T + 1, :])
    map_run_length = run_length_posterior.argmax(axis=1)
    new_run_prob = run_length_posterior[:, :cp_window].sum(axis=1)

    return BocpdResult(
        run_length_posterior=run_length_posterior,
        map_run_length=map_run_length,
        new_run_prob=new_run_prob,
        predictive_mean=predictive_mean,
        log_evidence=log_evidence,
        cp_window=cp_window,
    )


def _logsumexp(a: np.ndarray) -> float:
    a = np.asarray(a)
    m = a.max()
    if not np.isfinite(m):
        return m
    return float(m + np.log(np.sum(np.exp(a - m))))


def detect_changepoints_from_map(
    map_run_length: np.ndarray,
    min_drop: int = 10,
    refractory: int = 10,
) -> list[int]:
    """Locate change-points by detecting MAP run-length resets.

    During a stationary run BOCPD's MAP run-length grows by ~1 per step.
    A change-point manifests as a sharp drop (Δ < 0) once the posterior
    commits to the new regime. We report the FIRST index of each such
    drop and enforce a `refractory` window to avoid double-counting.
    """
    map_rl = np.asarray(map_run_length)
    drops = np.flatnonzero(np.diff(map_rl) < -min_drop) + 1
    out: list[int] = []
    last = -refractory - 1
    for d in drops:
        if d - last >= refractory:
            out.append(int(d))
            last = d
    return out
